src/input_stream.rs

```diff
@@ -341,6 +351,221 @@ impl<T: IsA<InputStream> + IsA<Seekable>> io::Seek for InputStreamRead<T> {
     }
 }
 
+enum State {
+    Waiting {
+        buffer: Vec<u8>,
+    },
+    Transitioning,
+    Reading {
+        pending: Pin<
+            Box<
+                dyn std::future::Future<Output = Result<(Vec<u8>, usize), (Vec<u8>, glib::Error)>>
+                    + 'static,
+            >,
+        >,
+    },
+    HasData {
+        buffer: Vec<u8>,
+        valid: (usize, usize), // both indexes are inclusives
+    },
+    Failed(crate::IOErrorEnum),
+}
+
+impl State {
+    fn into_buffer(self) -> Vec<u8> {
+        match self {
+            State::Waiting { buffer } => buffer,
+            _ => panic!("Invalid state"),
+        }
+    }
+
+    fn get_pending(
+        &mut self,
+    ) -> &mut Pin<
+        Box<
+            dyn std::future::Future<Output = Result<(Vec<u8>, usize), (Vec<u8>, glib::Error)>>
+                + 'static,
+        >,
+    > {
+        match self {
+            State::Reading { ref mut pending } => pending,
+            _ => panic!("Invalid state"),
+        }
+    }
+}
+pub struct InputStreamAsyncBufRead<T: IsA<InputStream>> {
+    stream: T,
+    state: State,
+}
+
+impl<T: IsA<InputStream>> InputStreamAsyncBufRead<T> {
+    pub fn into_input_stream(self) -> T {
+        self.stream
+    }
+
+    pub fn input_stream(&self) -> &T {
+        &self.stream
+    }
+
+    fn new(stream: T, buffer_size: usize) -> Self {
+        let buffer = vec![0; buffer_size];
+
+        Self {
+            stream,
+            state: State::Waiting { buffer },
+        }
+    }
+    fn set_reading(
+        &mut self,
+    ) -> &mut Pin<
+        Box<
+            dyn std::future::Future<Output = Result<(Vec<u8>, usize), (Vec<u8>, glib::Error)>>
+                + 'static,
+        >,
+    > {
+        match self.state {
+            State::Waiting { .. } => {
+                let waiting = mem::replace(&mut self.state, State::Transitioning);
+                let buffer = waiting.into_buffer();
+                let pending = self
+                    .input_stream()
+                    .read_async_future(buffer, Priority::default());
+                self.state = State::Reading { pending };
+            }
+            State::Reading { .. } => {}
+            _ => panic!("Invalid state"),
+        };
+
+        self.state.get_pending()
+    }
+
+    fn get_data(&self) -> Poll<io::Result<&[u8]>> {
+        if let State::HasData {
+            ref buffer,
+            valid: (i, j),
+        } = self.state
+        {
+            return Poll::Ready(Ok(&buffer[i..j + 1]));
```

Comment:
Why make the second one inclusive if you need to add +1 because of that here?

Reply:
Indeed, that makes arithmetic simpler.